spec/services/process_refund_service_spec.rb

```diff
@@ -0,0 +1,61 @@
+# frozen_string_literal: true
+
+require "rails_helper"
+
+RSpec.describe ProcessRefundService do
+  describe ".run" do
+    let(:finance_details) { double(:finance_details) }
+    let(:payment) { double(:payment, order_key: "123", registration_reference: "registration_reference", amount: 1_500) }
+    let(:user) { double(:user, email: "user@example.com") }
+    let(:payments) { double(:payments) }
+    let(:refund) { double(:refund) }
+
+    before do
+      expect(finance_details).to receive(:payments).and_return(payments)
+      expect(payments).to receive(:<<).with(refund)
+      expect(finance_details).to receive(:update_balance)
+      expect(finance_details).to receive(:save!)
+
+      expect(WasteCarriersEngine::Payment).to receive(:new).with(payment_type: WasteCarriersEngine::Payment::REFUND).and_return(refund)
+      expect(refund).to receive(:order_key=).with("123_REFUNDED")
+      expect(refund).to receive(:date_entered=).with(Date.current)
+      expect(refund).to receive(:date_received=).with(Date.current)
+      expect(refund).to receive(:amount=).with(-1_500)
+      expect(refund).to receive(:registration_reference=).with("registration_reference")
+      expect(refund).to receive(:updated_by_user=).with("user@example.com")
+      expect(payment).to receive(:worldpay?).and_return(worldpay)
+    end
+
+    context "when the payment is a card payment" do
+      let(:worldpay) { true }
+
+      it "generates a new refund payment and associate it with the right finance details" do
```

Comment:
Maybe this is my foggy post-holiday brain, but where are we testing that the payment is successfully generated and assigned to the finance details? Same question for the cash context.

Reply:
Yes we are :) is just all mocks and stubbing since this is an integration test. Most of the test code is in the `before` context